pasture-core/src/gpu/device.rs

```diff
@@ -0,0 +1,639 @@
+use wgpu::util::{DeviceExt, BufferInitDescriptor};
+use wgpu::BufferDescriptor;
+use crate::layout::{PointAttributeDataType};
+use crate::layout;
+use crate::containers::{PointBuffer};
+use bytemuck::__core::convert::TryInto;
+
+pub struct Device {
+    // Private fields
+    adapter: wgpu::Adapter,
+    device: wgpu::Device,
+    queue: wgpu::Queue,
+
+    upload_buffers: Vec<wgpu::Buffer>,
+    download_buffers: Vec<wgpu::Buffer>,
+    buffer_sizes: Vec<wgpu::BufferAddress>,
+    buffer_bindings: Vec<u32>,
+
+    cs_module: Option<wgpu::ShaderModule>,
+    bind_group: Option<wgpu::BindGroup>,
+    compute_pipeline: Option<wgpu::ComputePipeline>,
+}
+
+impl Device {
+    /// Create a device with default options:
+    /// - Low power GPU
+    /// - Primary backend for wgpu to use [Vulkan, Metal, Dx12, Browser]
+    pub async fn default() -> Device {
+        Device::new(DeviceOptions::default()).await
+    }
+
+    /// Create a device respecting the desired [DeviceOptions]
+    pub async fn new(device_options: DeviceOptions) -> Device {
+        // == Create an instance from the desired backend =========================================
+
+        let backend_bits = match device_options.device_backend {
+            DeviceBackend::Primary => { wgpu::BackendBit::PRIMARY }
+            DeviceBackend::Secondary => { wgpu::BackendBit::SECONDARY }
+            DeviceBackend::Vulkan => { wgpu::BackendBit::VULKAN }
+            DeviceBackend::Metal => { wgpu::BackendBit::METAL }
+            DeviceBackend::Dx12 => { wgpu::BackendBit::DX12 }
+            DeviceBackend::Dx11 => { wgpu::BackendBit::DX11 }
+            DeviceBackend::OpenGL => { wgpu::BackendBit::GL }
+            DeviceBackend::Browser => { wgpu::BackendBit::BROWSER_WEBGPU }
+        };
+
+        let instance = wgpu::Instance::new(backend_bits);
+
+        // == Create an adapter with the desired power preference =================================
+
+        let power_pref = if matches!(device_options.device_power, DevicePower::Low) {
+            wgpu::PowerPreference::LowPower
+        }
+        else {
+            wgpu::PowerPreference::HighPerformance
+        };
+
+        // The adapter gives us a handle to the actual device.
+        // We can query some GPU information, such as the device name, its type (discrete vs integrated)
+        // or the backend that is being used.
+        let adapter = instance.request_adapter(
+            &wgpu::RequestAdapterOptions {
+                power_preference: power_pref,
+                compatible_surface: None
+            }
+        ).await.unwrap();
```

Comment:
`unwrap()` should not be used unless we know that something can't fail. To me it seems reasonable to have the `new()` function return a `Result<Device>`, instead of a `Device`.